llvm/lib/Target/RISCV/RISCVISelLowering.cpp

```diff
@@ -13821,6 +13822,56 @@ static SDValue performCONCAT_VECTORSCombine(SDNode *N, SelectionDAG &DAG,
     Align = std::min(Align, Ld->getAlign());
   }
 
+  // If the load ptrs can be decomposed into a common (Base + Index) with a
+  // common constant stride, then return the constant stride. This matcher
+  // enables some additional optimization since BaseIndexOffset is capable of
+  // decomposing the load ptrs to (add (add Base, Index), Stride) instead of
+  // (add LastPtr, Stride) or (add NextPtr, Stride) that matchForwardStrided and
+  // matchReverseStrided use respectively.
+  auto matchConstantStride = [&DAG, &N](ArrayRef<SDUse> Loads) {
+    // Initialize match constraints based on the first load. Initialize
+    // ConstStride by taking the difference between the offset of the first two
+    // loads.
+    if (Loads.size() < 2)
+      return SDValue();
+    BaseIndexOffset BaseLdBIO =
```

Comment:
Hm, an idea...

Looking at this code, I realized we could factor out a helper routine of the form optional<int64_t> getPtrDiff(LD1, LD2).  Then we could loop over all of the indices, and check that the offset is known and the same.

After that, it occurred to me that both of the previous two matchers can *also* be written as rules inside a getPtrDiff helper.  (The return value has to become SDValue instead, but otherwise it works.)

Maybe we should switch over in a separate commit, and then implement this matcher in that style?

Reply:
> I realized we could factor out a helper routine of the form optional<int64_t> getPtrDiff(LD1, LD2). Then we could loop over all of the indices, and check that the offset is known and the same.

This seems to be very close to what BaseIndexOffset match does since if the (Base+Index) matches then the ptr diff is Offset. We could definitely add this helper method to simplify our checks in matchConstantStride.

 > After that, it occurred to me that both of the previous two matchers can also be written as rules inside a getPtrDiff helper. (The return value has to become SDValue instead, but otherwise it works.)

What happens in the case where there is no constant stride in matchFoward and matchReverse (i.e. stride can be a common add SDValue that is not a SDConstant).

Reply:
I think I understand, `getPtrDiff` would return the canonicalized representation of the non-constant value. We wouldn't be able to use BaseIndexOffset here. Do you have any ideas on what we could use?

Reply:
I pulled out getPtrDiff myself when trying wrap my head around this, hope this doesn't step on anyones toes: #69068
We should be able to extend it to check BaseIndexOffsets and return `DAG.getConstant` like we do here if they match

Reply:
@lukel97 thanks for your refactor. This change is very simple with getPtrDiff. I've updated this patch to be stacked on top of https://github.com/llvm/llvm-project/pull/69068